pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,666 @@
+
+                    debug!(?after, "disk usage");
+
+                    if after.has_pressure() {
+                        // Don't bother doing an out-of-order iteration here now.
+                        // In practice, the task period is set to a value in the tens-of-seconds range,
+                        // which will cause another iteration to happen soon enough.
+                        // TODO: deltas between the three different usages would be helpful,
+                        // consider MiB, GiB, TiB
+                        warn!(?outcome, ?after, "disk usage still high");
+                    } else {
+                        info!(?outcome, ?after, "disk usage pressure relieved");
+                    }
+                }
+            }
+        }
+        Err(e) => {
+            error!("disk_usage_eviction_iteration failed: {:#}", e);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Serialize)]
+#[allow(clippy::large_enum_variant)]
+pub enum IterationOutcome<U> {
+    NoPressure,
+    Cancelled,
+    Finished(IterationOutcomeFinished<U>),
+}
+
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[allow(dead_code)]
+#[derive(Debug, Serialize)]
+pub struct IterationOutcomeFinished<U> {
+    /// The actual usage observed before we started the iteration.
+    before: U,
+    /// The expected value for `after`, according to internal accounting, after phase 1.
+    planned: PlannedUsage<U>,
+    /// The outcome of phase 2, where we actually do the evictions.
+    ///
+    /// If all layers that phase 1 planned to evict _can_ actually get evicted, this will
+    /// be the same as `planned`.
+    assumed: AssumedUsage<U>,
+}
+
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[derive(Debug, Serialize)]
+#[allow(dead_code)]
+struct AssumedUsage<U> {
+    /// The expected value for `after`, after phase 2.
+    projected_after: U,
+    /// The layers we failed to evict during phase 2.
+    failed: LayerCount,
+}
+
+// The `#[allow(dead_code)]` is to suppress warnings about only the Debug impl reading these fields.
+// We use the Debug impl for logging, so, it's allright.
+#[allow(dead_code)]
+#[derive(Debug, Serialize)]
+struct PlannedUsage<U> {
+    respecting_tenant_min_resident_size: U,
+    fallback_to_global_lru: Option<U>,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Default, Serialize)]
+struct LayerCount {
+    file_sizes: u64,
+    count: usize,
+}
+
+#[allow(clippy::needless_late_init)]
+pub async fn disk_usage_eviction_task_iteration_impl<U: Usage>(
+    storage: &GenericRemoteStorage,
+    usage_pre: U,
+    cancel: &CancellationToken,
+) -> anyhow::Result<IterationOutcome<U>> {
+    static MUTEX: once_cell::sync::Lazy<tokio::sync::Mutex<()>> =
```

Comment:
I think the original idea was to limit http and background task from both doing the run at the same time. Do you have an idea how to accomplish this, moving up the stack?

Reply:
Yep, this is common for several background activity tasks we have. Gc can be both triggered by an API request and is run on a periodic basis. This part is common across these tasks. 

I'm not insisting on changing this right now, but in general for how I see it take a look on my review comment: https://github.com/neondatabase/neon/pull/3809#pullrequestreview-1359153995

Reply:
> I think the original idea was to limit http and background task from both doing the run at the same time.

... because originally, we didn't plan to expose the HTTP handler outside of tests.

We don't strictly need the mutex, I think.
But without it, with two concurrent `disk_usage_eviction_task_iteration_impl` calls, the outcome will be that the sets of evictees will overlap, so, one call will do the eviction and the other call will cry `warn!` in the log.

Reply:
Also I think this is related to tenant lifecycle management story. When tenant owns its tasks instead of task mgr it it'll be even more natural to have some common code for  that

Reply:
https://github.com/neondatabase/neon/pull/3887

Reply:
> Do we really need async mutex here? The only usage is synchronous try_lock.

Note on the mutex: https://github.com/neondatabase/neon/pull/3887/files#r1149561413

For our purposes it's a dropguarded atomic counter already, because it's based on 1-permit semaphore. Just no need to reimplement it. It does however have some extra we don't need. Compared to adding `defer`s I think it's a much better alternative to be used here, and probably in general in the sense of RAII vs. defer.

Reply:
I still disagree with async mutex usage. Async mutex is general is more expensive compared to synchronous one:

https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use

Namely: 
> This makes the async mutex more expensive than the blocking mutex, so the blocking mutex should be preferred in the cases where it can be used.

Reply:
> `std::sync::Mutex`

This doesn't work because of the `!Sync` on MutexGuard.

Negotiated off-github a comment line: b1d54024e705fbbcda7eee9cbc0b793fae439385